### PR TITLE
Refactor API

### PR DESCRIPTION
**Pull request by jamestalmage**
This is preliminary work on simplifying the API. The major (breaking) change is a new `test-run` event that passes an emitter for that specific test run. Said emitter is an instance of the new TestData class (might be more appropriately named TestRun). This contains test results that we were storing directly on the API instance. That really doesn't make sense anymore given that the API instance is called multiple times by Watcher. I feel this way makes a little more sense. There is definitely room for further refactoring, but this represents a big diff, and I want to get it reviewed and merged before further refactoring.

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -38,332 +32,156 @@ function Api(options) {
...
-	if (this.hasExclusive && !stats.hasExclusive) {
-		return;
-	}
+	var emitter = fork(file, options);
 
-	if (!this.hasExclusive && stats.hasExclusive) {
-		this.hasExclusive = true;
-		this.testCount = 0;
-	}
+	testData.listenToTestRun(emitter);
```
Comment (novemberborn) [inline api.js]: `observeFork`?

File: test/api.js [hunk 2]
```diff
@@ -6,6 +6,10 @@ var rimraf = require('rimraf');
 var test = require('tap').test;
 var Api = require('../api');
 var testDoublerPlugin = require('./fixture/babel-plugin-test-doubler');
+// var only = test;
+// if (process.env.ONLY) {
+// 	test = function () {};
+// }
```
Comment (novemberborn) [inline test/api.js]: ?

File: test/watcher.js [hunk 3]
```diff
@@ -431,7 +447,8 @@ group('chokidar is installed', function (beforeEach, test, group) {
 		t.plan(2);
 
 		files = ['foo-{bar,baz}.js'];
-		api.excludePatterns = ['!*bar*'];
+		avaFiles.defaultExcludePatterns.reset();
```
Comment (novemberborn) [inline test/watcher.js]: No need to reset, you can just use `.returns` to change the value.

File: test/watcher.js [hunk 4]
```diff
@@ -102,6 +102,22 @@ group('chokidar is installed', function (beforeEach, test, group) {
 		logger.finish.reset();
 		logger.reset.reset();
 
+		avaFiles.reset();
+		avaFiles.defaultExcludePatterns.reset();
+		avaFiles.defaultIncludePatterns.reset();
```
Comment (novemberborn) [inline test/watcher.js]: https://github.com/sindresorhus/ava/pull/713/files#r58517908 notwithstanding, it is indeed good practice to reset stub state in the `beforeEach` :smile:

**Comment by novemberborn** [comment]
> This contains test results that we were storing directly on the API instance. That really doesn't make sense anymore given that the API instance is called multiple times by Watcher. I feel this way makes a little more sense.

Yea that's always stuck me as strange.

Note though that the reporters reference these properties. Their tests use a stubbed `api` so they're still passing.

**Comment by novemberborn** [comment]
> Said emitter is an instance of the new TestData class (might be more appropriately named TestRun).

`RunStatus`?

File: test/api.js (hunk 2 above)
Comment (jamestalmage) [inline test/api.js]: Tap does not have exclusive tests. That was my hacky way of getting them

Comment (novemberborn) [inline test/api.js]: Neat trick!

**Comment by jamestalmage** [comment]
I set reporter.api to the new testData instance with each `test-run` event.

I added a to-do comment that that should be cleaned up. I might just change it to pass `testData` as a param to reporter events. Some reporters could be stateless that way.

**Comment by novemberborn** [comment]
> I might just change it to pass testData as a param to reporter events. Some reporters could be stateless that way.

:+1:

File: test/watcher.js [hunk 5]
```diff
@@ -431,7 +447,8 @@ group('chokidar is installed', function (beforeEach, test, group) {
 		t.plan(2);
 
 		files = ['foo-{bar,baz}.js'];
-		api.excludePatterns = ['!*bar*'];
+		// TODO(@jamestalmage, @novemberborn): There is no way for users to actually set exclude patterns yet.
+		avaFiles.defaultExcludePatterns.returns(['!*bar*']);
```
Comment (jamestalmage) [inline test/watcher.js]: @novemberborn - Currently, there is no way for users to set the exclude patterns. Is this a feature that needs to be exposed? 

You access `api.excludePatterns` in watcher, but that's a static value that never changes (I guess an API consumer could, but we haven't published that, so ...).

Should we open an issue about this, or is `avaFiles.defaultExcludePatterns()` all you need.

Comment (novemberborn) [inline test/watcher.js]: @jamestalmage you could add exclude patterns to `files`. We should just make sure to still use the default include patterns if `files` only contains negated patterns. That's how the watcher behaves with `source` patterns.

Comment (novemberborn) [inline test/watcher.js]: https://github.com/sindresorhus/ava/issues/736

File: lib/run-status.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+var path = require('path');
+var util = require('util');
+var chalk = require('chalk');
+var isObj = require('is-obj');
+var flatten = require('arr-flatten');
+var figures = require('figures');
+var EventEmitter = require('events').EventEmitter;
+var formatter = require('./enhance-assert').formatter();
+
+function TestData(opts) {
```
Comment (novemberborn) [inline lib/run-status.js]: @jamestalmage did you mean to rename this to `RunStatus`?

Comment (jamestalmage) [inline lib/run-status.js]: I did indeed

**Pull request merged**